plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java

```diff
@@ -0,0 +1,157 @@
+import org.eclipse.che.ide.api.action.ActionManager;
+import org.eclipse.che.ide.api.action.Presentation;
+import org.eclipse.che.ide.api.editor.annotation.QueryAnnotationsEvent;
+import org.eclipse.che.ide.api.editor.codeassist.CodeAssistCallback;
+import org.eclipse.che.ide.api.editor.codeassist.Completion;
+import org.eclipse.che.ide.api.editor.codeassist.CompletionProposal;
+import org.eclipse.che.ide.api.editor.document.Document;
+import org.eclipse.che.ide.api.editor.quickfix.QuickAssistInvocationContext;
+import org.eclipse.che.ide.api.editor.quickfix.QuickAssistProcessor;
+import org.eclipse.che.ide.api.editor.text.LinearRange;
+import org.eclipse.che.ide.api.editor.text.Position;
+import org.eclipse.che.ide.api.editor.text.TextPosition;
+import org.eclipse.che.ide.api.editor.text.annotation.Annotation;
+import org.eclipse.che.ide.api.icon.Icon;
+import org.eclipse.che.ide.api.parts.PerspectiveManager;
+import org.eclipse.che.plugin.languageserver.ide.editor.DiagnosticAnnotation;
+import org.eclipse.che.plugin.languageserver.ide.service.TextDocumentServiceClient;
+
+import com.google.gwt.user.client.rpc.AsyncCallback;
+import com.google.gwt.user.client.ui.Widget;
+
+/**
+ * A {@link QuickAssistProcessor} that implements LSP code actions as quick assists.
+ * @author Thomas Mäder
+ *
+ */
+public class LanguageServerQuickAssistProcessor implements QuickAssistProcessor {
+	private TextDocumentServiceClient textDocumentService;
+	private ActionManager actionManager;
+	private PerspectiveManager perspectiveManager;
+
+	@Inject
+	public LanguageServerQuickAssistProcessor(TextDocumentServiceClient textDocumentService, ActionManager actionManager, PerspectiveManager perspectiveManager) {
+		this.textDocumentService = textDocumentService;
+		this.actionManager = actionManager;
+		this.perspectiveManager= perspectiveManager;
+	}
+
+	@Override
+	public void computeQuickAssistProposals(QuickAssistInvocationContext invocationContext, CodeAssistCallback callback) {
+		LinearRange range = invocationContext.getTextEditor().getSelectedLinearRange();
+		Document document = invocationContext.getTextEditor().getDocument();
+		QueryAnnotationsEvent.AnnotationFilter filter = new QueryAnnotationsEvent.AnnotationFilter() {
+
+			@Override
+			public boolean accept(Annotation annotation) {
+				return annotation instanceof DiagnosticAnnotation;
+			}
+
+		};
+		QueryAnnotationsEvent.QueryCallback annotationCallback = new QueryAnnotationsEvent.QueryCallback() {
+
+			@Override
+			public void respond(Map<Annotation, Position> annotations) {
+				List<DiagnosticDTO> diagnostics = new ArrayList<>();
+				// iteration with range never returns anything; need to filter ourselves.
+				// https://github.com/eclipse/che/issues/4338
+				annotations.entrySet().stream().filter((e)->e.getValue().overlapsWith(range.getStartOffset(), range.getLength())).forEach((e) -> diagnostics.add(((DiagnosticAnnotation) e.getKey()).getDiagnostic()));
```

Comment:
I believe this line is too complex for reading I'm wondering if we could replace it for something more simple? Something like this:
```
int offset = range.getStartOffset();
int length =  range.getLength();
List<DiagnosticDTO> diagnostics =
annotations.entrySet()
                    .stream()
                    .filter(it -> it.getValue().overlapsWith(offset, length))
                    .map(it -> Map.Entry::getKey)
                    .map(it -> (DiagnosticAnnotation)it)
                    .map(it -> DiagnosticAnnotation::getDiagnostic)
                    .collect(Collectors.toList());
```
What do you think?